Fixing logic for date validation and adding test for it

aguilarm-umd opened the pull request:
Hello!

I had found an issue with this library when trying to use the from or until parameters for my oai-pmh service. When given date's formatted with timestamps "%Y-%m-%dT%H:%M:%SZ" the valid_date method would initially try to parse the date as "%Y-%m-%d" and would error, catch it, and not test for the other case. I forked the repository and added my changes, and also included a test as well.

The code would error out before checking the second date format, now it should be able to catch the Type and Value errors and continue to check the other date format. If it gets to the end of the for loop without returning anything, it will raise the OAIErrorBadArgument exception.

Also, when I was initially setting up my virtual environment for this repo, I was encountering problems with lxml because the version used in this repo does not work with m1 macbooks, so I updated the requirements.txt file to the latest version of lxml, but I did not include it here, I'm just mentioning it.

aguilarm-umd committed:
Changing logic for date validation and adding test for it

The code would error out before checking the second date format,
now it should be able to catch the Type and Value errors and continue to
check the other date format. If it gets to the end of the for loop
without returning anything, it will raise the OAIErrorBadArgument
exception.

natecollins commented:
Thanks for catching this! I'd be happy to merge your change in.

There is one other change included in your PR looks to be always allowing finer granularity in filter requests. Per the [OAI spec 3.3.1](https://www.openarchives.org/OAI/openarchivesprotocol.html#Dates), requests with finer granularity than the repo should throw an error. I read that as needing to keep the `if` condition for including the finer granularity, but perhaps I'm misunderstanding it. Let me know what you think.

Also, thanks for the heads up on lxml. I'll get that updated to 4.9 shortly.

aguilarm-umd commented:
Oh I didn't realize that's why the check was there, I can add that back. 

The only thing that seems a bit confusing to me on that spec is that I couldn't tell if there's granularities other than YYYY-MM-DD and YYYY-MM-DDThh:mm:ssZ. 

Like on wikipedia for the ISO 8601 they have an example with an offset, 2023‐09‐20T11:22:49−07:00, and I'm wondering if it's the case that a repo could be supporting that. 

And if that's the case, it seems that the library is right now hardcoded to support the above two cases.

aguilarm-umd commented:
Okay reading up a bit more I think its just the two formats that are supported, so I'll go and add that check back in

aguilarm-umd committed:
Re-adding back granularity check

Also adding parameter to the datasets class to allow formats with
timestamps for testing. And also modified a case to be a datastamp
with no timestamp.

natecollins commented:
Thanks, Marc!
